### PR TITLE
MTM-43100 providesDeviceRegistration fragment support for microservic…

### PR DESCRIPTION
**Pull request by Oezge-Akin**
Added "extensions" support in legacy applications' root level fragment.
This will be used to support extensible device registration feature: https://docs.google.com/document/d/1kdfPyqgpuWBYvrDu0eXYTt5-nFoyRWgamIaBYTEs6aI/edit#heading=h.ai877ax88i92
For kubernetes applications, it will be automatically added to application's manifest fragment already.

Other related PR:
Core model change: https://github.softwareag.com/IOTA/cumulocity-core/pull/339 (this needs to be merged first)
Lwm2m-agent change: https://github.softwareag.com/IOTA/cumulocity-agents/pull/132

### CODE REVIEW

**Review by chbakar — DISMISSED** [review]

**Review by ghost — DISMISSED** [review]

**Review by torsSAG — DISMISSED** [review]

File: microservice/subscription/src/main/java/com/cumulocity/microservice/subscription/annotation/EnableMicroserviceSubscriptionConfiguration.java [hunk 1]
```diff
@@ -87,6 +90,9 @@ public MicroserviceMetadataRepresentation metadata(Environment environment) thro
                 return MicroserviceMetadataRepresentation.microserviceMetadataRepresentation()
                         .requiredRoles(MoreObjects.firstNonNull(manifest.getRequiredRoles(), ImmutableList.<String>of()))
                         .roles(MoreObjects.firstNonNull(manifest.getRoles(), ImmutableList.<String>of()))
+                        .extensibleDeviceRegistration(
```
Comment (jaroslawherod) [inline microservice/subscription/src/main/java/com/cumulocity/microservice/subscription/annotation/EnableMicroserviceSubscriptionConfiguration.java]: Maybe you should extend ManifestRepresentation ?

Comment (jaroslawherod) [reply microservice/subscription/src/main/java/com/cumulocity/microservice/subscription/annotation/EnableMicroserviceSubscriptionConfiguration.java]: BTW Microservice manifest already have a concept of com.cumulocity.model.application.microservice.Extension
Maybe you can extend the representation to support it ?

**Review by jaroslawherod — CHANGES_REQUESTED** [review]
I think we should not extend the manifest it's already have extension concept that fits the requirements. 
Sorry if I've missed that before.

File: microservice/subscription/src/main/java/com/cumulocity/microservice/subscription/annotation/EnableMicroserviceSubscriptionConfiguration.java (hunk 1 above)
Comment (Oezge-Akin) [reply microservice/subscription/src/main/java/com/cumulocity/microservice/subscription/annotation/EnableMicroserviceSubscriptionConfiguration.java]: thanks, I will check this

File: microservice/subscription/src/main/java/com/cumulocity/microservice/subscription/model/application/ExtensibleDeviceRegistrationRepresentation.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        }
+
+        ExtensibleDeviceRegistrationRepresentation extensibleDeviceRegistration = new ExtensibleDeviceRegistrationRepresentation();
+        if (extensibleDeviceRegistrationMap.containsKey("name")) {
+            extensibleDeviceRegistration.setName((String) extensibleDeviceRegistrationMap.get("name"));
+        }
+        if (extensibleDeviceRegistrationMap.containsKey("description")) {
+            extensibleDeviceRegistration.setDescription((String) extensibleDeviceRegistrationMap.get("description"));
+        }
+        return  extensibleDeviceRegistration;
```
Comment (ghost) [inline microservice/subscription/src/main/java/com/cumulocity/microservice/subscription/model/application/ExtensibleDeviceRegistrationRepresentation.java]: Code formatting (extra white space)

Comment (Oezge-Akin) [reply microservice/subscription/src/main/java/com/cumulocity/microservice/subscription/model/application/ExtensibleDeviceRegistrationRepresentation.java]: class is removed with last commit.

File: microservice/subscription/src/main/java/com/cumulocity/microservice/subscription/annotation/EnableMicroserviceSubscriptionConfiguration.java (hunk 1 above)
Comment (Oezge-Akin) [reply microservice/subscription/src/main/java/com/cumulocity/microservice/subscription/annotation/EnableMicroserviceSubscriptionConfiguration.java]: now using `List<ExtensionRepresentation> extensions` in MicroserviceMetadataRepresentation and also extended the MicroserviceManifestRepresentation from core with this PR: https://github.softwareag.com/IOTA/cumulocity-core/pull/339

File: microservice/subscription/src/main/java/com/cumulocity/microservice/subscription/annotation/EnableMicroserviceSubscriptionConfiguration.java [hunk 3]
```diff
@@ -87,6 +88,7 @@ public MicroserviceMetadataRepresentation metadata(Environment environment) thro
                 return MicroserviceMetadataRepresentation.microserviceMetadataRepresentation()
                         .requiredRoles(MoreObjects.firstNonNull(manifest.getRequiredRoles(), ImmutableList.<String>of()))
                         .roles(MoreObjects.firstNonNull(manifest.getRoles(), ImmutableList.<String>of()))
+                        .extensions((List) manifest.getExtensions())
```
Comment (ghost) [inline microservice/subscription/src/main/java/com/cumulocity/microservice/subscription/annotation/EnableMicroserviceSubscriptionConfiguration.java]: why do we have casting here?

File: microservice/subscription/src/test/java/com/cumulocity/microservice/subscription/repository/impl/LegacyMicroserviceRepositoryTest.java [hunk 4]
```diff
@@ -320,6 +322,86 @@ public void shouldGetSubscriptions(){
...
+                    @Override
+                    public boolean matches(Object value) {
+                        if (value instanceof ApplicationRepresentation) {
+                            assertThat(((ApplicationRepresentation) value).get(MicroserviceMetadataRepresentation.EXTENSIONS_FIELD_NAME)).isNotNull();
+                            return true;
+                        }
+                        return false;
+                    }
+                });
+        assertThat(platform.take(byMethod(PUT))).isEmpty();
```
Comment (ghost) [inline microservice/subscription/src/test/java/com/cumulocity/microservice/subscription/repository/impl/LegacyMicroserviceRepositoryTest.java:335]: I don't really get how this test is working, the first `repository.register` is recording manifest with extensions only for the next call?
Can you make is more readable by extracting methods like `givenManifestWithExtensions` and maybe it make sense to split this one test into few different cases?

File: microservice/subscription/src/main/java/com/cumulocity/microservice/subscription/annotation/EnableMicroserviceSubscriptionConfiguration.java (hunk 3 above)
Comment (Oezge-Akin) [reply microservice/subscription/src/main/java/com/cumulocity/microservice/subscription/annotation/EnableMicroserviceSubscriptionConfiguration.java]: will remove :+1:

File: microservice/subscription/src/test/java/com/cumulocity/microservice/subscription/repository/impl/LegacyMicroserviceRepositoryTest.java (hunk 4 above)
Comment (Oezge-Akin) [reply microservice/subscription/src/test/java/com/cumulocity/microservice/subscription/repository/impl/LegacyMicroserviceRepositoryTest.java:335]: I am checking here create and update case so doing another register call after this and FakeCredentialsSwitchingPlatform platform used in the test class is able to manage POST and PUT correctly. 

I can still split it and do more readable :+1:

File: microservice/subscription/src/main/java/com/cumulocity/microservice/subscription/annotation/EnableMicroserviceSubscriptionConfiguration.java [hunk 5]
```diff
@@ -32,6 +32,7 @@
 import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.util.List;
```
Comment (c8y-kh4mell) [inline microservice/subscription/src/main/java/com/cumulocity/microservice/subscription/annotation/EnableMicroserviceSubscriptionConfiguration.java]: unsued

File: microservice/subscription/src/test/java/com/cumulocity/microservice/subscription/repository/impl/LegacyMicroserviceRepositoryTest.java [hunk 6]
```diff
@@ -320,6 +322,102 @@ public void shouldGetSubscriptions(){
...
 
+    @Test
+    public void shouldCreateAppWithExtensionsIfMetadataContains() {
+        repository.register(
+                getMetadataWithExtensions());
+
+        assertThat(platform.take(byMethod(POST)))
+                .hasSize(1)
+                .extracting("body")
+                .have(new Condition<Object>() {
```
Comment (c8y-kh4mell) [inline microservice/subscription/src/test/java/com/cumulocity/microservice/subscription/repository/impl/LegacyMicroserviceRepositoryTest.java]: condition could be extracted to self describing method in order to improve readibility

File: microservice/subscription/src/test/java/com/cumulocity/microservice/subscription/repository/impl/LegacyMicroserviceRepositoryTest.java [hunk 7]
```diff
@@ -320,6 +322,102 @@ public void shouldGetSubscriptions(){
                 .isSameAs(user);
     }
 
+    @Test
+    public void shouldCreateAppWithExtensionsIfMetadataContains() {
+        repository.register(
```
Comment (c8y-kh4mell) [inline microservice/subscription/src/test/java/com/cumulocity/microservice/subscription/repository/impl/LegacyMicroserviceRepositoryTest.java]: trivial (and optional): --> givenMicroserviceMetadataWithExtensions()

File: microservice/subscription/src/test/java/com/cumulocity/microservice/subscription/repository/impl/LegacyMicroserviceRepositoryTest.java [hunk 8]
```diff
@@ -320,6 +322,102 @@ public void shouldGetSubscriptions(){
...
+                .name(CURRENT_APPLICATION_NAME)
+                .build();
+        givenApplications(existing);
+
+        repository.register(getMetadataWithExtensions());
+
+        assertThat(platform.take(byMethod(PUT)))
+                .hasSize(1)
+                .extracting("body")
+                .have(new Condition<Object>() {
```
Comment (c8y-kh4mell) [inline microservice/subscription/src/test/java/com/cumulocity/microservice/subscription/repository/impl/LegacyMicroserviceRepositoryTest.java]: this is exactly the same condition as above, another reason to extract it to a method

**Review by c8y-kh4mell — DISMISSED** [review]

**Review by c8y-kh4mell — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Review by jaroslawherod — APPROVED** [review]

**Pull request merged**